gemini/gradio-voice/app.py

```diff
@@ -0,0 +1,253 @@
+import asyncio
+import base64
+import json
+import os
+from threading import Event
+
+from dotenv import load_dotenv
+import gradio as gr
+from gradio_webrtc import StreamHandler, WebRTC
+import numpy as np
+import websockets.sync.client
+
+load_dotenv()
+
+# mypy: disable-error-code="error-code-1, error-code-2"
+
+class GeminiConfig:
+    def __init__(self, api_key: str) -> None:
+        self.api_key = api_key
+        self.host = "generativelanguage.googleapis.com"
+        self.model = "models/gemini-2.0-flash-exp"
+        self.ws_url = f"wss://{self.host}/ws/google.ai.generativelanguage.v1alpha.GenerativeService.BidiGenerateContent?key={self.api_key}"
+
+
+class AudioProcessor:
+    @staticmethod
+    def encode_audio(data: np.ndarray, sample_rate: int) -> None:
+        """Encode Audio data to send to the server"""
+        encoded = base64.b64encode(data.tobytes()).decode("UTF-8")
+        return {
+            "realtimeInput": {
+                "mediaChunks": [
+                    {
+                        "mimeType": f"audio/pcm;rate={sample_rate}",
+                        "data": encoded,
+                    }
+                ],
+            },
+        }
+
+    @staticmethod
+    def process_audio_response(data: str) -> np.ndarray:
+        """Decode Audio data received from the server"""
+        audio_data = base64.b64decode(data)
+        return np.frombuffer(audio_data, dtype=np.int16)
+
+
+class GeminiHandler(StreamHandler):
+    def __init__(
+        self, expected_layout="mono", output_sample_rate=24000, output_frame_size=480
+    ) -> None:
+        super().__init__(
+            expected_layout,
+            output_sample_rate,
+            output_frame_size,
+            input_sample_rate=24000,
+        )
+        self.config = None
+        self.ws = None
+        self.all_output_data = None
+        self.audio_processor = AudioProcessor()
+        self.args_set = Event()
+
+    def copy(self) -> "GeminiHandler":
+        return GeminiHandler(
+            expected_layout=self.expected_layout,
+            output_sample_rate=self.output_sample_rate,
+            output_frame_size=self.output_frame_size,
+        )
+
+    def _initialize_websocket(self) -> None:
+        assert self.config, "Config not set"
+        try:
+            self.ws = websockets.sync.client.connect(self.config.ws_url, timeout=30)
+            initial_request = {
+                "setup": {
+                    "model": self.config.model,
+                }
+            }
+            self.ws.send(json.dumps(initial_request))
+            setup_response = json.loads(self.ws.recv())
+            print(f"Setup response: {setup_response}")
+        except websockets.exceptions.WebSocketException as e:
+            print(f"WebSocket connection failed: {str(e)}")
+            self.ws = None
```

Comment:
Can you try using the `genai.live.start_stream` method instead of raw websockets calls?

https://googleapis.github.io/python-genai/genai.html#module-genai.live